p2p/local.go

```diff
@@ -0,0 +1,124 @@
+package p2p
+
+import (
+	"context"
+	"time"
+
+	manet "gx/ipfs/QmRK2LxanhK2gZq6k6R7vk5ZoYZk8ULSSTB7FzDsMUX6CB/go-multiaddr-net"
+	ma "gx/ipfs/QmWWQ2Txc2c6tqjsBpzg5Ar652cHPGNsQQp2SejkNmkUMb/go-multiaddr"
+	pstore "gx/ipfs/QmXauCuJzmzapetmC6W4TuDJLL1yFFrVzSHoWv8YdbmnxH/go-libp2p-peerstore"
+	net "gx/ipfs/QmXfkENeeBvh3zYA51MaSdGUdBjhQ99cP5WQe8zgr6wchG/go-libp2p-net"
+	protocol "gx/ipfs/QmZNkThpqfVXs9GNbexPrfBbXSLNYeKrE7jwFM2oqHbyqN/go-libp2p-protocol"
+	peer "gx/ipfs/QmZoWKhxUmZ2seW4BzX6fJkNR8hh9PsGModr7q171yq2SS/go-libp2p-peer"
+)
+
+// localListener manet streams and proxies them to libp2p services
+type localListener struct {
+	ctx context.Context
+
+	p2p *P2P
+	id  peer.ID
+
+	proto string
+	laddr ma.Multiaddr
+	peer  peer.ID
+
+	listener manet.Listener
+}
+
+// ForwardLocal creates new P2P stream to a remote listener
+func (p2p *P2P) ForwardLocal(ctx context.Context, peer peer.ID, proto string, bindAddr ma.Multiaddr) (Listener, error) {
+	listener := &localListener{
+		ctx: ctx,
+
+		p2p: p2p,
+		id:  p2p.identity,
+
+		proto: proto,
+		laddr: bindAddr,
+		peer:  peer,
+	}
+
+	if err := p2p.Listeners.lock(listener); err != nil {
+		return nil, err
+	}
+
+	maListener, err := manet.Listen(bindAddr)
+	if err != nil {
+		p2p.Listeners.unlock()
+		return nil, err
+	}
+
+	listener.listener = maListener
+
+	p2p.Listeners.Register(listener)
+	go listener.acceptConns()
+
+	return listener, nil
+}
+
+func (l *localListener) dial() (net.Stream, error) {
+	ctx, cancel := context.WithTimeout(l.ctx, time.Second*30) //TODO: configurable?
+	defer cancel()
+
+	err := l.p2p.peerHost.Connect(ctx, pstore.PeerInfo{ID: l.peer})
+	if err != nil {
+		return nil, err
+	}
+
+	return l.p2p.peerHost.NewStream(l.ctx, l.peer, protocol.ID(l.proto))
+}
+
+func (l *localListener) acceptConns() {
```

Comment:
Not in this PR, but there's opportunity for logging errors inside of here. Maybe later if a p2p log-subsystem is added.